File: ide-features-trainer/src/training/lesson/Lesson.java
package training.lesson;

import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.ToggleAction;
import com.intellij.openapi.util.IconLoader;
import org.jetbrains.annotations.Nullable;
import training.editor.eduUI.EduIcons;
import training.lesson.exceptons.BadCourseException;
import training.lesson.exceptons.BadLessonException;

import java.awt.*;
import java.io.IOException;
import java.util.ArrayList;
import java.util.concurrent.ExecutionException;

/**
 * Created by karashevich on 29/01/15.
 */
public class Lesson extends AnAction {

    private Scenario scn;
    private String name;
    private String targetPath;
    private ArrayList<LessonListener> lessonListeners;
    private Course parentCourse;

    @Nullable
    private boolean isPassed;
    private boolean isOpen;

    public Lesson(Scenario scenario, boolean passed, @Nullable Course course) throws BadLessonException {

        super(scenario.getName());
            scn = scenario;
            name = scn.getName();

            isPassed = passed;
            if (!scn.getSubtype().equals("aimless")) {
                targetPath = scn.getTarget();
            } else {
                targetPath = null;
            }
            lessonListeners = new ArrayList<LessonListener>();
            parentCourse = course;

            isOpen = false;


    }

    public void open(Dimension infoPanelDimension) throws IOException, FontFormatException, LessonIsOpenedException {
        //init infoPanel, check that Lesson has not opened yet
        if (isOpen) throw new LessonIsOpenedException(this.getId() + "is opened");
        onStart();

        isOpen = true;
    }

    public void close(){
        //destroy infoPanel (infoPanel = null)
        isOpen = false;
        onClose();
    }

    /**
     *
     * @return null if DetailPanel still not initialized
     */
    @Nullable

    public String getId() {
        return name;
    }

    public boolean isPassed(){
        return isPassed;
    }

    public boolean isOpen() {return isOpen;}

    public void setPassed(boolean passed){
        isPassed = passed;
    }

    public Scenario getScn(){
        return scn;
    }

    @Nullable
    public String getTargetPath() {
        return targetPath;
    }

    @Nullable
    public Course getCourse() {return parentCourse;}

    //Listeners
    public void addLessonListener(LessonListener lessonListener){
        if (lessonListeners == null) lessonListeners = new ArrayList<LessonListener>();

        lessonListeners.add(lessonListener);
    }

    public void removeLessonListener(LessonListener lessonListener) {
        if (lessonListeners.contains(lessonListener)) {
            lessonListeners.remove(lessonListener);
        }
    }

    public void onStart(){
        if (lessonListeners == null) lessonListeners = new ArrayList<LessonListener>();

        for (LessonListener lessonListener : lessonListeners) {
            lessonListener.lessonStarted(this);
        }
    }

    public void onClose(){
        for (LessonListener lessonListener : lessonListeners) {
            lessonListener.lessonClosed(this);
        }

        lessonListeners = null;

    }

    //call onPass handlers in lessonListeners
    public void onPass(){
        if(isPassed) return;
        scn.getRoot().getAttribute("passed").setValue("true");
        try {
            scn.saveState();
            for (LessonListener lessonListener : lessonListeners) {
                lessonListener.lessonPassed(this);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        isPassed = true;
    }

    public void onNextLesson() throws BadLessonException, ExecutionException, IOException, FontFormatException, InterruptedException, BadCourseException, LessonIsOpenedException {
        for (LessonListener lessonListener : lessonListeners) {
            lessonListener.lessonNext(this);
        }
    }


//    @Override
//    public boolean isSelected(AnActionEvent anActionEvent) {
//        return isPassed;
//    }
//
//    @Override
//    public void (AnActionEvent anActionEvent, boolean b) {
//        try {
//            CourseManager.getInstance().openLesson(anActionEvent.getProject(), this);
//        } catch (BadCourseException e) {
//            e.printStackTrace();
//        } catch (BadLessonException e) {
//            e.printStackTrace();
//        } catch (IOException e) {
//            e.printStackTrace();
//        } catch (FontFormatException e) {
//            e.printStackTrace();
//        } catch (InterruptedException e) {
//            e.printStackTrace();
//        } catch (ExecutionException e) {
//            e.printStackTrace();
//        } catch (LessonIsOpenedException e) {
//            e.printStackTrace();
//        }
//
//    }


    @Override
    public void actionPerformed(AnActionEvent anActionEvent) {
        try {
            CourseManager.getInstance().openLesson(anActionEvent.getProject(), this);
        } catch (BadCourseException e) {
            e.printStackTrace();
        } catch (BadLessonException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } catch (FontFormatException e) {
            e.printStackTrace();
        } catch (InterruptedException e) {
            e.printStackTrace();
        } catch (ExecutionException e) {
            e.printStackTrace();
        } catch (LessonIsOpenedException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void update(AnActionEvent e) {
        if (isPassed())
            e.getPresentation().setIcon(IconLoader.getIcon(EduIcons.CHECKMARK_DARK_GRAY));

    }
}

File: ide-features-trainer/src/training/learn/lesson/Lesson.kt
package training.learn.lesson

import com.intellij.openapi.project.Project
import training.learn.CourseManager
import training.learn.Module
import training.learn.exceptons.BadLessonException
import training.learn.exceptons.BadModuleException
import training.learn.exceptons.LessonIsOpenedException
import training.learn.exceptons.NoProjectException
import training.learn.log.LessonLog
import java.awt.Dimension
import java.awt.FontFormatException
import java.io.IOException
import java.util.*
import java.util.concurrent.ExecutionException

/**
 * Created by karashevich on 29/01/15.
 */
data class Lesson(val scenario: Scenario, val lang: String, val module: Module?) {

  var lessonListeners: ArrayList<LessonListener> = ArrayList<LessonListener>()
    private set
  var statistic = ArrayList<Pair<String, Long>>()
  var exerciseCount: Short = 0
    private set
  var passed: Boolean = false
  var isOpen: Boolean = false
    private set
  val name: String = scenario.name
  val id: String = scenario.id

  /*Log lesson metrics*/
  val lessonLog: LessonLog = LessonLog(this)

  init {
    lessonListeners = ArrayList<LessonListener>()
    passed = LessonStateManager.getStateFromBase(id) == LessonState.PASSED
  }

  @Deprecated("")
  @Throws(IOException::class, FontFormatException::class, LessonIsOpenedException::class)
  fun open(infoPanelDimension: Dimension) {
    //init infoPanel, check that Lesson has not opened yet
    if (isOpen) throw LessonIsOpenedException(this.name + "is opened")
    onStart()
    isOpen = true
  }


  @Throws(NoProjectException::class, BadLessonException::class, ExecutionException::class, LessonIsOpenedException::class, IOException::class, FontFormatException::class, InterruptedException::class, BadModuleException::class)
  fun open() {
    var currentProject = CourseManager.instance.currentProject
    if (currentProject == null) {
      currentProject = CourseManager.instance.learnProject
    }
    if (currentProject == null) throw NoProjectException()
    CourseManager.instance.openLesson(currentProject, this)
  }

  @Throws(NoProjectException::class, BadLessonException::class, ExecutionException::class, LessonIsOpenedException::class, IOException::class, FontFormatException::class, InterruptedException::class, BadModuleException::class)
  fun open(projectWhereToOpenLesson: Project) {
    CourseManager.instance.openLesson(projectWhereToOpenLesson, this)
  }

  fun close() {
    isOpen = false
    onClose()
  }

  //Listeners
  fun addLessonListener(lessonListener: LessonListener) {
    lessonListeners.add(lessonListener)
  }

  fun removeLessonListener(lessonListener: LessonListener) {
    if (lessonListeners.contains(lessonListener)) lessonListeners.remove(lessonListener)

  }

  fun onStart() {
    lessonLog.log("Lesson started")
    exerciseCount = 0
    statistic.add(Pair("started", System.currentTimeMillis()))
    lessonLog.resetCounter()
    lessonListeners.forEach { it.lessonStarted(this) }
  }

  private fun onItemPassed() {
    statistic.add(Pair("passed item #" + exerciseCount, System.currentTimeMillis()))
    exerciseCount++
  }

  private fun onClose() {
    lessonListeners.clear()
    statistic.add(Pair("closed", System.currentTimeMillis()))

  }

  //call onPass handlers in lessonListeners
  private fun onPass() {
    lessonLog.log("Lesson passed")
    statistic.add(Pair("finished", System.currentTimeMillis()))
    lessonListeners.forEach { it.lessonPassed(this) }

  }

  @Throws(BadLessonException::class, ExecutionException::class, IOException::class, FontFormatException::class, InterruptedException::class, BadModuleException::class, LessonIsOpenedException::class)
  fun onNextLesson() {
    lessonListeners.forEach { it.lessonNext(this) }
  }

  fun pass() {
    passed = true
    LessonStateManager.setPassed(this)
    onPass()
  }

  fun passItem() {
    onItemPassed()
  }

  internal object EditorParameters {
    val PROJECT_TREE = "projectTree"
  }

  override fun equals(other: Any?): Boolean {
    return other != null && other is Lesson && other.name == this.name
  }

  override fun hashCode(): Int {
    var result = scenario.hashCode()
    result = 31 * result + id.hashCode()
    result = 31 * result + name.hashCode()
    return result
  }

}
